library(dplyr)
library(pathview)
library(data.table)
library(tibble)
library(org.Dr.eg.db)

## This script reads the *_diffData.txt file generated by DESeq2 pipeline script
## Maps the significant DEGs on the given KEGG pathways


rm(list = ls())

path <- "E:/Chris_UM/Analysis/CoreData/15_ZhuBo_RNASeq2/PG_PV_diff"
setwd(path)

##Prepare input data
compare <- c("PG", "PV")

sampleInfoFile <- "sampleInfo.txt"
diffFile <- "PG_vs_PV_diffData.txt"
pathFile <- "pathwayOfInterest.txt"


outFilePrefix <- paste0(compare, collapse = "_vs_")
p_cutoff <- 0.05
lfc_cut <- 1
up_cut <- lfc_cut
down_cut <- lfc_cut * -1


exptInfo <- read.table(file = sampleInfoFile, header = T, sep = "\t", row.names = "sampleID")
pathways <- read.table(file = pathFile, header = T, sep = "\t", stringsAsFactors = F)

## if required, modify the row names
rownames(exptInfo) <- sub("_WT", "_", rownames(exptInfo))

diffData <- fread(diffFile, sep = "\t", stringsAsFactors = F, header = T, data.table = F, na.strings = c('NA'))
degGenes <- filter(diffData, padj < p_cutoff, log2FoldChange >= up_cut | log2FoldChange <= down_cut) %>%
  column_to_rownames(var = "geneID")




storePath = paste0(c(path, "pathway"), collapse = "/")
setwd(storePath)

## required by Pathview for organism dre 
data(gene.idtype.bods)

pv.out <- pathview(gene.data = degGenes["log2FoldChange"], pathway.id = pathways$pathID, 
                   species = "dre", out.suffix = "pathview", kegg.native = T,
                   gene.idtype = "ENSEMBL", gene.annotpkg = "org.Dr.eg.db",
                   na.col = "transparent", limit=list(gene=3, cpd=3), bins = list(gene=20, cpd=20),
                   low = "blue", mid = "grey", high = "red",
                   same.layer = T,
                   new.signature = F)

######################################################################################
######################################################################################







